Add websocket support for port forwarding

fraenkel opened the pull request:
#32880

**Release note**:
```release-note
Port forwarding can forward over websockets or SPDY.
```

sebgoa commented:
@ngtuna can you build a "release" with this patch, we will test it with Helm.

ngtuna commented:
@sebgoa OK Seb.

On Tuesday, October 4, 2016, sebgoa <notifications@github.com> wrote:

> @ngtuna <https://github.com/ngtuna> can you build a "release" with this
> patch, we will test it with Helm.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/33684#issuecomment-251310744>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AB31x0UF4t0v9y3UEQgpE3VTpAwp48KLks5qwfufgaJpZM4KJOYi>
> .
>


-- 
Regards,
Tu Nguyen
Sent from mobile

sttts commented on pkg/kubelet/server/portforward/portforward.go:
```diff
@@ -0,0 +1,78 @@
+)
+
+const (
+	NonZeroExitCodeReason = unversioned.StatusReason("NonZeroExitCode")
```
are these used anywhere?

fraenkel replied:
Sorry, it was a copy paste error. I will remove it.

fraenkel commented on pkg/kubelet/server/portforward/websocket.go:
```diff
@@ -0,0 +1,143 @@
+	}, nil
+}
+
+func createWebSocketStreams(req *http.Request, w http.ResponseWriter, portForwarder PortForwarder, podName string, uid types.UID, supportedPortForwardProtocols []string, idleTimeout, streamCreationTimeout time.Duration) (*context, error) {
```
It was to match how exec did it... I was trying to see if I could make websockets and spdy align like remoteexec did but no luck so far.

sttts commented on pkg/kubelet/server/portforward/portforward.go:
```diff
@@ -0,0 +1,53 @@
+		err = createHttpStreamStreams(req, w, portForwarder, podName, uid, supportedProtocols, idleTimeout, streamCreationTimeout)
+	}
+
+	if err != nil {
```
It's strange that `createStreams` actually blocks and handles the connection.

sttts commented on pkg/kubelet/server/portforward/httpstream.go:
```diff
@@ -0,0 +1,309 @@
+// forward streams. It checks each stream's port and stream type headers,
+// rejecting any streams that with missing or invalid values. Each valid
+// stream is sent to the streams channel.
+func portForwardStreamReceived(streams chan httpstream.Stream) func(httpstream.Stream, <-chan struct{}) error {
```
as we are in the `portforwarding` package here, we can strip off the `portForward` from the identifiers. `streamReceivedHandler` would fit better here.

sttts commented on pkg/kubelet/server/portforward/httpstream.go:
```diff
@@ -0,0 +1,309 @@
+
+// portForwardStreamHandler is capable of processing multiple port forward
+// requests over a single httpstream.Connection.
+type portForwardStreamHandler struct {
```
same here, no `portForward` prefix necessary.

sttts replied:
In fact, for concistency this should be called `httpSteamHandler`

fraenkel commented on pkg/kubelet/server/portforward/httpstream.go:
```diff
@@ -0,0 +1,309 @@
+		uid:       uid,
+		forwarder: portForwarder,
+	}
+	h.run()
```
I will rename the method. This was originally trying to make websockets and spdy work as one but in the end its not possible given the differences. the creates will now be handleXXXX

sttts commented:
@fraenkel it would be much easier to review if this PR would consist of two commits: one for moving the old http handler code, a second for adding the WS code.

fraenkel committed:
Port forward over websockets

- split out port forwarding into its own package

Allow multiple port forwarding ports
- Make it easy to determine which port is tied to which channel
- odd channels are for data
- even channels are for errors

- allow comma separated ports to specify multiple ports

Add  portfowardtester 1.2 to whitelist

fraenkel commented on pkg/kubelet/server/portforward/httpstream.go:
```diff
@@ -0,0 +1,309 @@
+/*
```
Not really, it originally started as portforward.go but portforward.go was reimplemented so you won't see it as a move. I could rename it in the original commit so it doesn't change as much but it does get affected because of ServePortForward.

sttts commented on pkg/kubelet/server/portforward/websocket.go:
```diff
@@ -0,0 +1,143 @@
+}
+
+// newOptions creates a new options from the Request.
+func newOptions(req *http.Request) (*options, error) {
```
as we have introduced protocol version v4 here, the newOptions should be versioned as well, same for the `option struct` as well.

sttts commented on pkg/kubelet/server/portforward/websocket.go:
```diff
@@ -0,0 +1,143 @@
+
+// newOptions creates a new options from the Request.
+func newOptions(req *http.Request) (*options, error) {
+	portString := req.FormValue(api.PortHeader)
```
Could we support multiple-port multiplexing by using a comma separated port list here?

fraenkel replied:
we could but the spdy code does it as separate requests. Not as the initial request.

sttts replied:
Sounds good. If comma separated lists work with some additional work, I feel much better about the limited, implemented design.

fraenkel replied:
We have a choice, either we do a comma separated list or we allow multiple port headers. Either would accomplish this but we do have to choose.

sttts replied:
That's why I asked about why form values are used and not headers. For headers, as you say, there are multiple values and comma separated lists in each (if I understand the RFC correctly: https://tools.ietf.org/html/rfc2616#section-4.2).

sttts commented on pkg/kubelet/server/portforward/websocket.go:
```diff
@@ -0,0 +1,143 @@
+		forwarder: portForwarder,
+	}
+
+	h.portForward()
```
For http it's called `h.run()`.

fraenkel commented:
With the addition of multiple ports, I am now considering adding the port number in the initial flow from the server to the client when the "channel" is created.  All odd channels would be data and all even channels are error. Right now it's relying on port order which can work but flowing it removes a bit of the guess.

sttts commented on pkg/kubelet/server/portforward/websocket.go:
```diff
@@ -43,25 +44,33 @@ const (
+		if len(portString) == 0 {
+			return nil, fmt.Errorf("%q is cannot be empty", api.PortHeader)
+		}
+		port, err := strconv.ParseUint(portString, 10, 16)
```
Compare https://tools.ietf.org/html/rfc2616#section-4.2. Comma separated lists must be supported as well.

fraenkel replied:
websockets cannot set headers. comma separate parameters for query params isn't defined but seems to be optionally supported. I can easily add that.

sttts commented:
- Can we have an e2e test for this?
- Can you update the PR description?

Otherwise, IMO it's ready from my side.

/cc @ncdc @smarterclayton ptal

ncdc commented:
I will add this to my review queue. I have a lot in front of it, so it might take a few days before I can look at it.

smarterclayton commented:
Same.

fraenkel commented:
@sttts I started to build out the e2e test but realized that I will end up building a standalone client to make this all work. The client side for this function is out of scope. Did we have some other way of testing this e2e?

fraenkel commented:
Resolved how we will test this but currently investigating why/where query params are dropped in the APIServer when forwarding to the kubelet.  Is there a reason why query parameters aren't forwarded?

smarterclayton commented:
What query parameters would you expect?  We only forward things we
guarantee we accept to prevent information disclosure and proxy attacks in
general.

On Thu, Oct 27, 2016 at 5:58 PM, Michael Fraenkel <notifications@github.com>
wrote:

> Resolved how we will test this but currently investigating why/where query
> params are dropped in the APIServer when forwarding to the kubelet. Is
> there a reason why query parameters aren't forwarded?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/33684#issuecomment-256781852>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABG_p8IAkUX5o1qFiC56sTX-SKHQxANGks5q4R5xgaJpZM4KJOYi>
> .
>

fraenkel commented:
Right now the ports to forward are sent via query parameters because websockets doesn't support headers.

fraenkel commented:
I see now that I need to tweak the pod strategy to allow the port parameter to flow for this case.

smarterclayton commented:
Ok.  Please add a test that verifies that *only* that parameter is passed
through by the strategy.

On Thu, Oct 27, 2016 at 6:26 PM, Michael Fraenkel <notifications@github.com>
wrote:

> I see now that I need to tweak the pod strategy to allow the port
> parameter to flow for this case.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/33684#issuecomment-256787474>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABG_pwdcd_WGQbu2fbhaAyfhaNvcJqerks5q4ST7gaJpZM4KJOYi>
> .
>

fraenkel commented:
I am trying to make []int32 my ports parameter but the openapi/swagger generation barfs over it. We noticed that for PodExecOptions, command is []string but the generated openapi converts it to a string and not type as an array nor allowedMultiple=true.
Is there some limitation on what you can specify as query parameters. I can deal with ports being a string or even []string, but it just seems wrong.

fraenkel commented:
We found out how []string was done and did the same, so making progress to get the e2e working again.

fraenkel commented:
@stts can we get this reviewed sooner than later?

sttts commented:
@fraenkel Where are we with the addressed comments? @smarterclayton had some. And the e2e test was open? Is that all fixed?

fraenkel commented:
All comments were addressed. The test regarding query params is done in the same style as remote command which is all indirect. There is a complete e2e test.

sttts commented on pkg/kubelet/server/portforward/websocket.go:
```diff
@@ -95,49 +107,77 @@ func handleWebSocketStreams(req *http.Request, w http.ResponseWriter, portForwar
-	defer h.streams[dataChannel].Close()
-	defer h.streams[errorChannel].Close()
-	defer h.conn.Close()
+	wg := sync.WaitGroup{}
```
the `conn.Close()` is gone here. Intentional?

fraenkel replied:
Yes. The streams are closed in portForward and the connection is closed above in handleWebSocketStreams.

sttts commented on pkg/kubelet/server/streaming/server.go:
```diff
@@ -89,7 +89,7 @@ type Config struct {
 	StreamIdleTimeout:     4 * time.Hour,
 	StreamCreationTimeout: remotecommand.DefaultStreamCreationTimeout,
-	SupportedProtocols:    remotecommand.SupportedStreamingProtocols,
+	SupportedProtocols:    append(remotecommand.SupportedStreamingProtocols, portforward.SupportedProtocols...),
```
is this correct to merge the protocols for remotecommand and portforward? Would have expected to have two separate lists.

sttts commented on pkg/kubelet/server/streaming/server.go:
```diff
@@ -312,7 +312,8 @@ func (s *server) servePortForward(req *restful.Request, resp *restful.Response)
 		s.config.StreamIdleTimeout,
-		s.config.StreamCreationTimeout)
+		s.config.StreamCreationTimeout,
+		s.config.SupportedProtocols)
```
here I would expect `s.config.SupportPortForwardProtocols`

sttts commented on pkg/registry/core/pod/strategy.go:
```diff
@@ -380,6 +380,14 @@ func streamParams(params url.Values, opts runtime.Object) error {
 			params.Add(api.ExecTTYParam, "1")
 		}
+	case *api.PodPortForwardOptions:
+		if len(opts.Ports) > 0 {
```
does it make sense to throw and error here if `Ports` is empty? We require at least one port in the server.

sttts commented on test/e2e/common/pods.go:
```diff
@@ -564,6 +604,127 @@ var _ = framework.KubeDescribe("Pods", func() {
+			return nil
+		}, time.Minute, 10*time.Second).Should(BeNil())
+
+		By("Sending the expected data to the local port")
```
lower case

sttts commented on test/e2e/common/pods.go:
```diff
@@ -564,6 +604,127 @@ var _ = framework.KubeDescribe("Pods", func() {
+			return nil
+		}, time.Minute, 10*time.Second).Should(BeNil())
+
+		By("Verifying logs")
```
lower case

sttts commented on test/e2e/common/pods.go:
```diff
@@ -564,6 +604,127 @@ var _ = framework.KubeDescribe("Pods", func() {
+		Eventually(func() error {
+			channel, msg, err := wsRead(ws)
+			if err != nil {
+				return fmt.Errorf("Failed to read completely from websocket %s: %v", url.String(), err)
```
errors are lower case, also below

sttts commented:
Some questions, some nits. Overall looks good!

fraenkel commented:
@stts PTAL, I addressed all of your comments and ended up introducing some new tests to validate the strategy.

sttts commented:
From my side it looks good. Was https://github.com/kubernetes/kubernetes/pull/33684#issuecomment-256788784 addressed?

sttts commented:
And can you create a release-note section in the PR description?

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/33684)
<!-- Reviewable:end -->

fraenkel commented:
@sttts The new test checks what the strategy creates. It generates the query from the PortForwardOption. Not sure how else to test it.  It is more than what remote command has.

sttts commented:
The new test is client side. Wondering whether something like `?ports=123&dangerous=true` should be rejected by the server. @smarterclayton if you are around, can you comment?

sttts commented:
@fraenkel can you squash into two commit: code changes + generated code. I would say this is ready to go.

sttts commented:
@k8s-bot test this

sttts commented:
@k8s-bot test this

fraenkel commented:
@stts am I missing something are all the environments screwed up when it comes to port forwarding? Or is there something else I need to do to make it all work?

sttts commented:
@k8s-bot test this

sttts commented:
@k8s-bot test this

sttts commented:
@k8s-bot ok to test

fraenkel commented:
@sttts 
CRI GCE E2E, the problem is the CRI layer. The code they added to kubelet/server/server.go is wrong for portForward especially now that there are query params that must flow.

GCE Node e2e passed. I cannot tell what failed since it isn't obvious.

For the rest, they all have the same problem, port forwarding for websockets works and SPDY fails. However the logs show websockets has failed. I will take a closer look. The logs show what looks like the container dieing and has socat getting a broken pipe. It looks like readiness isn't in place everywhere. Or we are being killed before we are setup. I don't see the readiness container ever completely start.

fraenkel committed:
CRI Portforward needs to forward websocket ports

- adjust ports to int32
- CRI flows the websocket ports as query params

- Do not validate ports since the protocol is unknown
  SPDY flows the ports as headers and websockets uses query params
- Only flow query params if there is at least one port query param

fraenkel commented:
@sttts Finally figured out the issues. PTAL at the new commits.
I had to remove one of the validations and let it occur at the kubelet when it knows the if its websockets or spdy.

sttts commented on pkg/registry/core/pod/strategy.go:
```diff
@@ -381,14 +381,13 @@ func streamParams(params url.Values, opts runtime.Object) error {
 			params.Add(api.ExecTTYParam, "1")
 		}
 	case *api.PodPortForwardOptions:
-		if len(opts.Ports) == 0 {
```
do we have this test somewhere else then?

sttts replied:
I mean will the connection eventually fail for len(ports) == 0?

fraenkel replied:
Yes it is ./pkg/kubelet/server/portforward/websocket.go:65. It fails when it is a websocket request and allows it for spdy. I am going to make a quick fix on the message, it isn't proper english.

sttts replied:
ok.

Congrats to finding the issue!

fraenkel commented:
What I have noticed:
unit/integration is a 50/50 due to the OpenAPI test which takes 5m. Those that pass take 35m and those that fail take 15m of total time across all PRs.

GCE Node e2e has a random issue as seen by the following:
Jan 07 22:03:46 localhost kubelet[2284]: E0107 22:03:46.509853    2284 websocket.go:184] error forwarding port 80 to pod pod-portforward-websocket-264be22d-d525-11e6-a51a-42010a800006_e2e-tests-pods-w25lr, uid : unable to do port forwarding: socat not found.

sttts commented:
socat not found?!

fraenkel commented:
Jan 08 13:47:10 localhost kubelet[2233]: E0108 13:47:10.084227    2233 websocket.go:184] error forwarding port 80 to pod pod-portforward-websocket-f0a6704d-d5a8-11e6-bb2d-42010a80000b_e2e-tests-pods-vnx28, uid : unable to do port forwarding: socat not found. /usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin

It should have found /usr/bin/socat.  Will add more debug.

fraenkel commented:
on GCE Node, the CoreOS alpha and Ubuntu trusty nodes do not have socat installed.

sttts commented:
How do the other portforwarding tests work?

fraenkel commented:
So I added tests to common/pods which runs on more node types, i.e., the standard tests in portforward do not run.
One option is to move my tests to portforward but we are just hiding a setup issue.

fraenkel commented:
Seems we have https://github.com/kubernetes/kubernetes/issues/32128 and https://github.com/kubernetes/kubernetes/issues/29213. Not sure how the ones in our tests are built.

sttts commented:
/cc @euank ^^

sttts commented:
@fraenkel I would be fine with moving them to portforward. The issue applies to the existing portwarding tests as well and is already hidden, isn't it?

fraenkel commented:
@sttts Seems something has changed. I can no longer generate code that makes the tests pass and not have a conflict. If I remove the above file, I get no conflict, all tests pass except verification which says the api-reference needs to be updated. I do that and this is what happens.

sttts commented on pkg/kubelet/server/server_test.go:
```diff
@@ -1528,11 +1529,11 @@ func TestServePortForward(t *testing.T) {
 			}
 
-			p, err := strconv.ParseUint(test.port, 10, 16)
+			p, err := strconv.ParseInt(test.port, 10, 32)
```
A last thing: why was this changed into 32 bits?

fraenkel replied:
I can change it back but it was to match how it was being carried in all the data structures. It seems we use int32 to refer to ports. CRI and others already had int32 everywhere.

The int32 usage was done by https://github.com/kubernetes/kubernetes/issues/17095. Ports had an option and they went int32.

sttts commented:
@derekwaynecarr would you mind to take a look at the kubelet part of this PR, especially everything around CRI?

smarterclayton commented:
Can you split the API changes into their own PR which we can merge first?  That'll make this easier.

smarterclayton commented:
Assign that PR to me.

fraenkel commented:
@smarterclayton Which API changes in particular? Everything under pkg/api?

fraenkel commented:
APIs are now under #39797

sttts commented:
APIs merged 🚀

fraenkel commented:
@sttts rebased and ready to go

smarterclayton commented:
A few questions

fraenkel commented:
@smarterclayton Just a comment to document the protocol under the handleWebSocketStreams. Let me know if I need to add more detail.

sttts commented:
Anything left here? @smarterclayton all your comments addressed?

fraenkel committed:
Generated code

smarterclayton commented on pkg/kubelet/server/portforward/websocket.go:
```diff
@@ -45,21 +45,25 @@ const (
 	portStrings := req.URL.Query()[api.PortHeader]
 	if len(portStrings) == 0 {
-		return nil, fmt.Errorf("%q is required", api.PortHeader)
+		return nil, fmt.Errorf("%q header is required", api.PortHeader)
```
query parameter, not header

fraenkel replied:
So the issue and I can fix it is that it is a header for SPDY and a query parameter for WebSockets. I can have 2 even though they are the same name.

smarterclayton commented on pkg/kubelet/server/portforward/websocket.go:
```diff
@@ -45,21 +45,25 @@ const (
 	for _, portString := range portStrings {
 		if len(portString) == 0 {
-			return nil, fmt.Errorf("%q is cannot be empty", api.PortHeader)
+			return nil, fmt.Errorf("%q cannot be empty", api.PortHeader)
```
`query parameter %q cannot be empty`

smarterclayton commented:
The image is built and pushed, right?

smarterclayton commented:
Just the last two minor comments and then this LGTM and @sttts can reapply labels as necessary.

smarterclayton commented:
/approve

smarterclayton commented:
@k8s-bot kops aws e2e test this

sttts commented:
@k8s-bot unit test this issue #38518

liggitt commented:
@fraenkel looks like this is [failing the CRI CI](https://k8s-testgrid.appspot.com/google-cri#gce&include-filter-by-regex=websocket). 

cc @timstclair

fraenkel commented:
@liggitt The error in the logs is when the upgrade to websockets fail. We have seen similar issues in other environments as flakes but usually an issue with a proxy. This error only occurs when the client attempts to set up websockets. I have never seen a complete end to end log which shows the pieces in the middle which is rejecting the websocket upgrade.

smarterclayton commented:
@fraenkel do you have an example of a client using this today?

fraenkel commented:
@smarterclayton Nope, the only example I know is the e2e test I wrote which just shows the mechanics behind the protocol. The stream handling should be similar for portforward and exec/proxy, its just the setup and hook up that is slightly different.
